Speed up DataTierAllocationDecider

original-brownbear opened the pull request:
This decider is quite slow in `allocationAllowed` and shows up in profiling.
We should be able to get a much tighter loop with this change that avoids building
the list of role names over and over and removes some dead code as well.
There's certainly additional speedups possible here though, this just deals with the most obvious issue.

original-brownbear committed:
Speed up DataTierAllocationDecider

This decider is quite slow in `allocationAllowed` and shows up in profiling.
We should be able to get a much tighter loop with this change that avoids building
the list of role names over and over and removes some dead code as well.

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

original-brownbear committed:
logic is hard

joegallo commented:
Do you happen to know how good are the tests here? I added an exception into `allocationAllowed` and I couldn't get any failing tests out of it -- that has me on edge.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into speed-up-data-tier-allocation-decider

original-brownbear committed:
incorporate review

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:
```diff
@@ -209,25 +210,29 @@ static boolean tierNodesPresent(String singleTier, DiscoveryNodes nodes) {
 
-    private static boolean allocationAllowed(OpType opType, String tierSetting, Set<DiscoveryNodeRole> roles) {
+    private static boolean allocationAllowed(String tierSetting, Set<DiscoveryNodeRole> roles) {
         String[] values = parseTierList(tierSetting);
```
We actually don't need to parse this setting at all, because the caller of this function does:

```java
            Optional<String> tier = preferredTierFunction.apply(tierPreference, allocation.nodes());
            if (tier.isPresent()) {
                String tierName = tier.get();
                if (allocationAllowed(tierName, roles)) {
```

They `preferredTierFunction` returns the optional value of a *single* tier name.

So we can guarantee that in this function, `tierSetting` is a non-empty, single-valued string for a given tier (that also exists in the cluster). So, all callers will go through the `values.length == 1` branch (assuming they aren't a generic `data` role node)

joegallo replied:
Ah, excellent! d4668a8c17c

dakrone reviewed (changes requested):
I left a couple of comments, I think we can simplify this more

joegallo committed:
Remove outdated comment

joegallo committed:
Internal refactor

The tierName argument is a single tier name, rather than comma
separated list of zero or more tier names.

joegallo commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:
```diff
@@ -160,7 +161,7 @@ private Decision shouldIndexPreferTier(IndexMetadata indexMetadata, Set<Discover
                 String tierName = tier.get();
                 // The OpType doesn't actually matter here, because we have
                 // selected only a single tier as our "preferred" tier
```
9ca47b0c449

joegallo commented:
Great catch, @dakrone, I've added two commits that should address your comments.

joegallo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into speed-up-data-tier-allocation-decider

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:
```diff
@@ -192,7 +185,12 @@ private Decision shouldIndexPreferTier(IndexMetadata indexMetadata, Set<Discover
 
     public static String[] parseTierList(String tiers) {
-        return Strings.tokenizeToStringArray(tiers, ",");
+        if (Strings.hasLength(tiers) == false) {
```
To avoid an array like `[" "]` (not that we expect that), maybe we should use:
```suggestion
        if (Strings.hasText(tiers) == false) {
```

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:
```diff
@@ -209,25 +207,19 @@ static boolean tierNodesPresent(String singleTier, DiscoveryNodes nodes) {
-        String[] values = parseTierList(tierSetting);
-        for (String value : values) {
+    private static boolean allocationAllowed(String tierName, Set<DiscoveryNodeRole> roles) {
+        assert Strings.hasLength(tierName) : "tierName must be not null and non-empty, but was [" + tierName + "]";
```
Same here, 
```suggestion
        assert Strings.hasText(tierName) : "tierName must be not null and non-empty, but was [" + tierName + "]";
```

dakrone reviewed (approved):
LGTM, I left two really minor comments

joegallo commented:
@elasticmachine run elasticsearch-ci/part-2

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into speed-up-data-tier-allocation-decider

original-brownbear committed:
CR: comments

original-brownbear commented:
Jenkins run elasticsearch-ci/part-2

original-brownbear commented:
Thanks Joe + Lee!
